keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+
+## Motivation
+
+Current logging in the Kubernetes control plane doesn’t guarantee any uniform structure for log messages and references to Kubernetes objects in those logs. This makes parsing, processing, storing, querying and analyzing logs hard and forces administrators and developers to rely on ad-hoc solutions in most cases based on some regular expressions. Due to those problems any analytical solution based on those logs is hard to implement and maintain.
+
+### Goals
+
+* Improve querability of most common logs in Kubernetes by standardizing log message and references to Kubernetes objects (Pods, Nodes etc.)
+* Enforce log structure by introduction of new klog methods that could be used to generate structured logs.
+* Propose reasonable scope of migration:
+  * Executable within SIG-instrumentation resources
+  * Minimize review burden on other SIGs
+  * Leave logging in acceptable state if effort fails
+* Simplify ingestion of logs into third party logging solutions by adding an option to output logs in the JSON format
+
+
+### Non-Goals
+
+* We are **not** replacing currently used logging library (klog) or the way in which it is used
+* We are **not** preserving full compatibility of logging output
+* We are **not** proposing structuring all logs in Kubernetes
+
+## Proposal
+
+This KEP proposes introducing a standardized log message structure. To enforce a new log message structure we are proposing adding new methods to the klog library: `InfoS`, `WarningS` , `ErrorS`  and `FatalS` which will provide a structured interface for building log messages and creating new helper methods which will provide consistent identification of Kubernetes objects in logs.
+
+With the new methods we would like also to migrate to klog v2 and introduce a new JSON logging output format to Kubernetes components that would be an alternative to the current text format and which will make querying and processing Kubernetes logs even simpler.
+
+Taking into account the size of Kubernetes repository we will not tackle the problem of full migration to new klog methods but instead focus on improving the querability of most common logs. We will be targeting reaching a high percentage of logs in a new format generated by real cluster.
+
+### Log message structure
+
+We would like for the Kubernetes community to settle on one preferred log message structure. Proposed structure should:
+* Separate log message from its arguments
+* Treat log arguments as key-value pairs
+* Be easily parsable and queryable
+* Have a specific guidance on log message and its arguments
+
+For this purpose we suggest to use following log message structure:
+```
+<message> <key1>=<value1> <key2>=<value2> ...
+```
+
+where
+* message is formatted using “%q” fmt logic
+* keys are formatted using “%s” fmt logic
+* values are formatted using “%v” fmt logic and optionally escaped using the “%q” fmt logic if it not a number or a boolean
+* keys are separated from values with a “=” sign
```

Comment:
Can keys contain an '=' ?

Reply:
Should we define a valid character set for keys?
If validation performance is an issue, should we have a debug mode that logs warnings for invalid keys?

Reply:
FTR: we've seen a bunch of issues in the past where glog/klog was consuming significant percentage (e.g. 20 or 50) of CPU usage of a component, for logs it wasn't even printing. The reason was that it is first computing the log (which e.g.can serialize some large objects) and only then check whether to print it or no based on the verbosity. 
If we fix this problem, I hope that we will be able to afford some validation.

Reply:
Keys should not include `=`. Additionally to runtime validation that produces warning we could introduce static analysis that detects invalid keys. Keys should be provided as static strings and could be verified via code analysis.